### PR TITLE
[diffusion, cfg, tests] feat: save rollout and validation videos as mp4

### PR DESCRIPTION
**Pull request by NancyFyong**
### What does this PR do?

Make the diffusion FlowGRPO trainer save **video** rollouts, not just images.
`_dump_generations` and `_maybe_log_val_generations` assumed image tensors
`[N, C, H, W]` and permuted them directly, which raised on 5-D video batches
`[N, T, C, H, W]`. This branches on tensor rank so video runs write one `.mp4`
per sample (via `diffusers.export_to_video`) while the image path stays
byte-for-byte, and wandb receives an encoded mp4 file path instead of raw
frames (avoiding the heavy `moviepy` dependency).

It also fixes a color-inversion bug: `diffusers.export_to_video` rescales NumPy
input by 255 even when it is already `uint8 [0, 255]`, overflowing to a
photographic negative (`255 - x`). Frames now go through
`verl_omni.utils.media.video_tensor_to_pil_frames`, which hands `export_to_video`
PIL images (not rescaled) and sanitizes non-finite decoder output before
quantization.

### Checklist Before Starting

- [x] Search for similar PRs:
  - https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+dump_generations (0 results)
  - https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+rollout+video
  - https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+diffusion+video+mp4 (0 results)

  No existing PR touches `_dump_generations` / `_maybe_log_val_generations` or the
  video-vs-image dump branching. Nearby diffusion PRs (#278 LTX2.3 t2av pipeline,
  #300 distillation losses, #172 rollout server routing) address unrelated scope.
- [x] PR title formatted as `[{modules}] {type}: {description}`.

### Test

All new tests are pure-CPU and run in seconds (`diffusers` + `imageio`/ffmpeg only).

```bash
python -m pytest -q \
  tests/utils/test_media_on_cpu.py \
  tests/trainer/diffusion/test_dump_generations_video_on_cpu.py \
  tests/trainer/diffusion/test_val_video_logging_on_cpu.py
# -> 10 passed
```

- `test_media_on_cpu.py` — exact uint8 pixel values, NaN/±inf/clip handling, shape rejection.
- `test_dump_generations_video_on_cpu.py` — one `.mp4` per video sample, one `.jpg` per image
  sample (unchanged), `max_samples` bound, and a decode-and-check that warm colors are **not**
  inverted (R > G > B, R > 128, B < 128 after ffmpeg round-trip).
- `test_val_video_logging_on_cpu.py` — monkeypatched `wandb.Video` gets a real, existing,
  imageio-decodable mp4 and the temp dir is cleaned up after `log()`; and the **real** `wandb.Video`
  ingests our production-encoded mp4 (matching byte size + sha256) with no `moviepy`.

`pre-commit run --files <changed files>` passes (ruff, ruff-format, mypy, generated-config
verification, license, etc.).

Manually verified end-to-end against a live wandb run: the encoded mp4 uploads and is stored
server-side as `media/videos/.../generations_0_<sha>.mp4` with summary `_type: video-file`,
which the wandb UI renders as an inline playable video.

### API and Usage Example

New, backward-compatible `trainer` config knobs (image path unchanged at defaults):

| knob | default | effect |
|---|---|---|
| `video_fps` | 24 | fps for encoded mp4s (train dump, val dump, wandb) |
| `rollout_data_save_freq` | 1 | dump train rollouts every N steps (`<=0` = never); bounds per-step ffmpeg cost |
| `rollout_data_max_samples` | null | cap number of train rollout samples dumped |
| `validation_data_max_samples` | null | cap number of validation samples dumped |

```yaml
trainer:
  logger: [console, wandb]
  log_val_generations: 4        # log up to 4 validation videos to wandb
  rollout_data_dir: /path/to/rollouts   # dump train rollout mp4s to disk
  validation_data_dir: /path/to/val     # dump validation mp4s to disk
  video_fps: 16
  rollout_data_save_freq: 10
  rollout_data_max_samples: 8
```

Coverage: on-disk `.mp4` for **both** train rollouts (`rollout_data_dir`) and validation
(`validation_data_dir`); wandb videos for **validation** (`log_val_generations > 0` + `wandb`
in `logger`). Train rollouts are dumped to disk only, matching the existing generations-logging
design.

### Design & Code Changes

- `verl_omni/utils/media.py` (new) — `video_tensor_to_pil_frames`: `[T, 3, H, W]` float tensor
  -> list of PIL RGB frames, with NaN/±inf sanitize + clamp + uint8 quantization. Shared by both
  dump and wandb paths so color handling is identical.
- `verl_omni/trainer/diffusion/ray_diffusion_trainer.py`
  - `_dump_generations`: branch on rank — 5-D -> `{i}.mp4` via `export_to_video`, 4-D -> `{i}.jpg`
    (unchanged). Honors `max_samples`.
  - `_maybe_log_val_generations`: pre-encode each per-sample video to a temp mp4 and pass
    `wandb.Video` the file path (images still `wandb.Image`); wrapping runs after
    select/shuffle/truncate so only retained samples are encoded; temp dir removed after `log()`.
  - `_log_rollout_data`: pass `video_fps` / `rollout_data_max_samples` through to `_dump_generations`.
- `verl_omni/trainer/config/diffusion_trainer.yaml` (+ regenerated `_generated_*` references) —
  add the four knobs above.
- Tests as listed under **Test**.

### Checklist Before Submitting

- [x] Read the Contribute Guide.
- [x] Apply pre-commit checks (passed on changed files).
- [ ] Add / Update the documentation. (New knobs documented in this PR description; no dedicated
      docs page for these trainer knobs today.)
- [x] Add unit test(s) covering the code (CPU tests above; a live GPU + wandb path is not feasible
      in CI, verified manually as noted).

---

AI assistance (Claude Code) was used for this change. A human author reviewed every changed line
and ran the tests above.

### CODE REVIEW

File: verl_omni/utils/media.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    prevents that second scaling and the resulting modulo-256 color inversion.
+    Values outside the display range, including non-finite decoder output, are
+    sanitized before quantization.
+    """
+    if video.ndim != 4 or video.shape[1] != 3:
+        raise ValueError(f"Expected an RGB video tensor with shape [T, 3, H, W], got {tuple(video.shape)}")
+
+    frames = []
+    for frame in video:
+        frame = frame.detach().permute(1, 2, 0).to(dtype=torch.float32)
```
Comment (SamitHuang) [inline verl_omni/utils/media.py]: why not directly permute on the video? so as the rest processing

**Review by SamitHuang — COMMENTED** [review]
Overall LGTM. Pls check `video_tensor_to_pil_frames` efficiency  @NancyFyong

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py [hunk 2]
```diff
@@ -275,35 +276,56 @@ def _create_dataloader(self, train_dataset, val_dataset, collate_fn, train_sampl
...
+    def _dump_generations(
+        self, inputs, outputs, gts, scores, reward_extra_infos_dict, dump_path, max_samples=None, fps=24
+    ):
+        """Dump rollout/validation samples to disk as media files plus a JSONL index.
+
+        ``outputs`` is a batch of either images ``[N, C, H, W]`` or videos
+        ``[N, T, C, H, W]``. Images are written as ``{i}.jpg`` and videos as ``{i}.mp4``
+        (at ``fps`` frames per second). ``max_samples`` caps how many samples are written
+        (``None`` = all); this bounds disk usage and the per-step ffmpeg encode cost that
+        would otherwise be prohibitive for video runs.
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/ray_diffusion_trainer.py]: clean

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py [hunk 3]
```diff
@@ -357,25 +381,61 @@ def _maybe_log_val_generations(self, inputs, outputs, scores):
...
-        if "wandb" in self.config.trainer.logger:
-            import wandb
+        import numpy as np
 
-            outputs = [wandb.Image(image.float(), file_type="jpg") for image in outputs]
-        samples = list(zip(inputs, outputs, scores, strict=True))
+        # Pair (input, output, score), sort by input text, then deterministically shuffle
+        # and truncate BEFORE wrapping media so video runs do not encode every sample just
+        # to keep a handful. Sorting keys on the input string only, so raw output tensors
+        # are never compared.
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/ray_diffusion_trainer.py]: clean or one line comment

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py [hunk 4]
```diff
@@ -357,25 +381,61 @@ def _maybe_log_val_generations(self, inputs, outputs, scores):
...
+        samples = list(zip(inputs, list(outputs), scores, strict=True))
         samples.sort(key=lambda x: x[0])  # Sort by input text
-
-        # Use fixed random seed for deterministic shuffling
-        rng = np.random.RandomState(42)
+        rng = np.random.RandomState(42)  # Fixed seed for deterministic shuffling
         rng.shuffle(samples)
-
-        # Take first N samples after shuffling
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/ray_diffusion_trainer.py:374]: keep old comments stays

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py [hunk 5]
```diff
@@ -357,25 +381,61 @@ def _maybe_log_val_generations(self, inputs, outputs, scores):
...
+
+        # Log to each configured logger. wandb copies each Video's file into the run
+        # directory during log(); the temp encodes are removed afterwards -- in a
+        # ``finally`` so a logging failure (network/wandb error) cannot leak the temp
+        # dir and let it accumulate over long runs.
+        try:
+            self.validation_generations_logger.log(self.config.trainer.logger, samples, self.global_steps)
+        finally:
+            if video_tmp_dir is not None:
+                shutil.rmtree(video_tmp_dir, ignore_errors=True)
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/ray_diffusion_trainer.py]: clean or one-line comment, pls follow our repo style

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py [hunk 6]
```diff
@@ -1120,9 +1182,12 @@ def fit(self):
                     actor_output_metrics = reduce_metrics(actor_output.meta_info["metrics"])
                     metrics.update(actor_output_metrics)
 
-                    # Log rollout generations if enabled
+                    # Log rollout generations if enabled. rollout_data_save_freq bounds how
+                    # often we dump (default 1 = every step); this matters for video runs
+                    # where per-step encoding is expensive.
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/ray_diffusion_trainer.py]: keep just `# Log rollout generations if enabled`

File: verl_omni/utils/media.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+
+def video_tensor_to_pil_frames(video: torch.Tensor) -> list[Image.Image]:
+    """Convert an RGB ``[T, C, H, W]`` tensor in ``[0, 1]`` to PIL frames.
+
+    PIL frames are intentional here. ``diffusers.utils.export_to_video`` scales
+    NumPy inputs by 255 even when they are already ``uint8``; passing PIL images
+    prevents that second scaling and the resulting modulo-256 color inversion.
+    Values outside the display range, including non-finite decoder output, are
+    sanitized before quantization.
```
Comment (zhtmike) [inline verl_omni/utils/media.py]: keep it clean

**Comment by zhtmike** [comment]
please clean AI comments, lot of them are self-explanatory and trivial , we don't need add them here.

**Comment by NancyFyong** [comment]
> please clean AI comments, lot of them are self-explanatory and trivial , we don't need add them here.

Thanks for review，perhaps could we provide skills for pr-style, code-style and so on？If needed, I can provide assistance and first list an RFC.

**Comment by zhtmike** [comment]
> > please clean AI comments, lot of them are self-explanatory and trivial , we don't need add them here.
> 
> Thanks for review，perhaps could we provide skills for pr-style, code-style and so on？If needed, I can provide assistance and first list an RFC.

yeah, very necessary! cc @AndyZhou952

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py [hunk 8]
```diff
@@ -374,8 +396,35 @@ def _maybe_log_val_generations(self, inputs, outputs, scores):
...
+                    if video_tmp_dir is None:
+                        video_tmp_dir = tempfile.mkdtemp(prefix="val_video_")
+                    frames = video_tensor_to_pil_frames(out)
+                    video_path = os.path.join(video_tmp_dir, f"{len(wrapped)}.mp4")
+                    export_to_video(frames, video_path, fps=fps)
+                    media = wandb.Video(video_path, format="mp4")
+                else:
+                    media = wandb.Image(out.float(), file_type="jpg")
+                wrapped.append((inp, media, score))
+            samples = wrapped
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/ray_diffusion_trainer.py]: it is better to write it into a helper function, like `verl/utils/tracking.py`, we can create our self version and put it there, with extra cpu test for this

File: verl_omni/utils/media.py
Comment (zhtmike) [inline verl_omni/utils/media.py:1]: there is similar things in verl_omni/utils/reward_score/reward_utils.py, should we put them together?

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py (hunk 8 above)
Comment (NancyFyong) [reply verl_omni/trainer/diffusion/ray_diffusion_trainer.py]: ok，I‘ll fix it

File: verl_omni/utils/media.py
Comment (NancyFyong) [reply verl_omni/utils/media.py:1]: sure, I've move to reward_utils

**Review by zhtmike — APPROVED** [review]
